### PR TITLE
IPROTO features (IPROTO_ID)

### LINKED ISSUE #6253: Add iproto request to query server features
Currently, the only way for a client (connector or other Tarantool instance) to check if the server supports a certain feature is to analyze the version string sent by the server in a greeting message, such as

`Tarantool 2.9.0 (Binary) bb381d4c-e6cb-4633-8422-98f52433011b`

This approach complicates client code and looks fragile. Besides, it doesn't allow us to change the version format reported in a greeting message (#6183) without taking a risk of breaking older clients.

The suggested solution is to add a new iproto request that would allow a client to query the feature set available on the server.

### PR DESCRIPTION
**Pull request by locker**
This patch set introduces a new IPROTO command (`IPROTO_ID`) for exchanging features between a server and a client. A client is supposed to issue the command right after establishing a connection to let the server know about features it supports. In reply to the command, the server sends its own feature set. Basing on the request/reply, server/client can enable or disable certain functionality.

Apart from adding the new command, this patch set also updates the net.box connector to support it. After it is applied, a client can specify features it expects to be available on the server in net.box options. If a required feature is unavailable, the connection will fail.

Currently, there are only two features available, `IPROTO_FEATURE_STREAMS` and `IPROTO_FEATURE_TRANSACTIONS`, both of which are always set both by the net.box connector and by the server. In the near future, we will add a new feature that will supersede the `error_marshaling_enabled` session setting: it will enable encoding errors as the `MP_ERROR` msgpack extension if the connector supports it, see #6428.

For more details, see comments to individual patches and the [design document](https://docs.google.com/document/d/1Fj3819bvMrxYHPILBZdvQW4quDw_RjhG4n9o85W0JxU).

Closes #6253

### CODE REVIEW

File: src/box/iproto_features.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "iproto_features.h"
+
+#include <msgpuck.h>
```
Comment (Gerold103) [inline src/box/iproto_features.c:8]: Could you please use `""` for non-system headers? Msgpuck here is not from the system, we always use the header stored in the source code.

File: src/box/iproto_features.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#include <stdint.h>
+
+struct iproto_features IPROTO_CURRENT_FEATURES;
+
+uint32_t
+mp_sizeof_iproto_features(struct iproto_features *features)
```
Comment (Gerold103) [inline src/box/iproto_features.c:14]: The argument could be `const`. The same in `mp_encode_iproto_features`.

File: src/box/iproto_features.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	size += mp_sizeof_array(count);
+	return size;
+}
+
+char *
+mp_encode_iproto_features(char *data, struct iproto_features *features)
+{
+	int count = 0;
+	iproto_features_foreach(features, feature_id)
+			count++;
```
Comment (Gerold103) [inline src/box/iproto_features.c:31]: Excess indentation.

File: src/box/iproto_features.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	 */
+	IPROTO_FEATURE_TRANSACTIONS = 1,
+	iproto_feature_id_MAX,
+};
+
+/**
+ * IPROTO protocol feature bit map.
+ */
+struct iproto_features {
+	char bits[BITMAP_SIZE(iproto_feature_id_MAX)];
```
Comment (Gerold103) [inline src/box/iproto_features.h:38]: IMO, would be fine to hardcode the size here, and add an assertion to `iproto_features_init` that `bitmap_size()` is the same as the hardcoded value. Then wouldn't need the first commit. But I am ok with this solution too.

File: src/box/xrow.c [hunk 5]
```diff
@@ -423,6 +424,38 @@ iproto_reply_ok(struct obuf *out, uint64_t sync, uint32_t schema_version)
...
+iproto_reply_id(struct obuf *out, uint64_t sync, uint32_t schema_version)
+{
+	size_t size = IPROTO_HEADER_LEN;
+	size += mp_sizeof_map(2);
+	size += mp_sizeof_uint(IPROTO_VERSION);
+	size += mp_sizeof_uint(IPROTO_CURRENT_VERSION);
+	size += mp_sizeof_uint(IPROTO_FEATURES);
+	size += mp_sizeof_iproto_features(&IPROTO_CURRENT_FEATURES);
+
+	char *buf = obuf_reserve(out, size);
```
Comment (Gerold103) [inline src/box/xrow.c:458]: If the size is exact, you can use `obuf_alloc()` right away.

File: src/box/iproto_features.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ */
+char *
+mp_encode_iproto_features(char *data, struct iproto_features *features);
+
+/**
+ * Decodes a IPROTO protocol features bit map from msgpack.
+ * Advances the data pointer. Returns true on success.
+ */
+bool
+mp_decode_iproto_features(const char **data, struct iproto_features *features);
```
Comment (Gerold103) [inline src/box/iproto_features.h:118]: Could you please use int as a result value? We usually use `0`/`-1` to detect errors.

File: changelogs/unreleased/gh-6253-net-box-iproto-id.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+## feature/core
```
Comment (Gerold103) [inline changelogs/unreleased/gh-6253-net-box-iproto-id.md:1]: For the commit message:
- `currently, both sets are empty, but this will change in the future` - shouldn't they be non-empty already now? According to the patch they are not empty. Netbox supports both streams and transactions. As well as the version of tarantool server you are working on.
- From the names `required_protocol_features` and `peer_protocol_features` I would drop the word 'protocol'. Up to you.
- `there are two features are defined` - one 'are' is excess.

File: changelogs/unreleased/gh-6253-net-box-iproto-id.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+
```
Comment (Gerold103) [inline changelogs/unreleased/gh-6253-net-box-iproto-id.md:2]: As for the feature about requesting a certain version and features. Now looking at the patch, I am not sure it is needed((. Netbox could just support everything what it can depending on the tarantool version and `IPROTO_ID`, and raise errors on attempts to use the not defined features. For example, if you try to use `netbox:begin()` and the protocol does not support transactions, you will get an error anyway. Either netbox could raise it right away, or the server will. What do you think?

Maybe add it if somebody asks? Otherwise it complicates the netbox API a bit.

File: src/box/lua/net_box.c [hunk 9]
```diff
@@ -58,6 +59,18 @@
...
+	/**
+	 * IPROTO protocol version supported by the netbox connector.
+	 */
+	NETBOX_IPROTO_VERSION = 0,
+};
+
+/**
+ * IPROTO protocol features supported by the netbox connector.
+ */
+static struct iproto_features NETBOX_IPROTO_FEATURES;
```
Comment (Gerold103) [inline src/box/lua/net_box.c:72]: Why is it in caps? The same for `IPROTO_CURRENT_FEATURES`. Don't we usually use caps for macros and enums? And in Lua for constants.

File: test/box/net.box_iproto_id_gh-6253.result [hunk 10]
```diff
@@ -0,0 +1,276 @@
+-- test-run result file version 2
```
Comment (Gerold103) [inline test/box/net.box_iproto_id_gh-6253.result:1]: For features AFAIK we do not specify issue number in the test file name. It applies only to bugfix tests.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: changelogs/unreleased/gh-6253-iproto-id.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+## feature/core
+
+* Extended the network protocol (IPROTO) with a new request type (`IPROTO_ID`)
+  that is supposed to be used for exchanging sets of supported features between
+  server and client (gh-6253).
```
[the remark this answers is not in the document]
Comment (locker) [reply changelogs/unreleased/gh-6253-iproto-id.md:5]: You're right, we don't really need them, but I'd like to have some non-empty initial feature set, because it's useful for testing and documentation. These two features were added recently (in beta) and they were major recent extension of the protocol so I think it should be fine to add them. I guess we could add `IPROTO_CHUNK` to the feature set as well.

File: src/box/iproto_features.c (hunk 1 above)
Comment (locker) [reply src/box/iproto_features.c:8]: Okay.

File: src/box/iproto_features.c (hunk 2 above)
Comment (locker) [reply src/box/iproto_features.c:14]: Fixed.

File: src/box/iproto_features.c (hunk 3 above)
Comment (locker) [reply src/box/iproto_features.c:31]: Fixed.

File: src/box/iproto_features.h (hunk 4 above)
Comment (locker) [reply src/box/iproto_features.h:38]: Using bitmap_size at compile time is more convenient IMO. Let's leave it as is.

File: src/box/iproto_features.h (hunk 6 above)
Comment (locker) [reply src/box/iproto_features.h:118]: Done.

File: src/box/lua/net_box.c (hunk 9 above)
Comment (locker) [reply src/box/lua/net_box.c:72]: This is to emphasize that they are constants. If I could, I'd make them macros. Plus it's consistent with `IPROTO_CURRENT_VERSION`.

File: test/box/net.box_iproto_id_gh-6253.result (hunk 10 above)
Comment (locker) [reply test/box/net.box_iproto_id_gh-6253.result:1]: I don't think there's such a rule. E.g. `test/box/net.box_timeout-gh-3107.test.lua` tests a feature (async timeout), but still appends a ticket number. I think a ticket number doesn't hurt so let's leave it as is.

File: changelogs/unreleased/gh-6253-net-box-iproto-id.md (hunk 7 above)
Comment (locker) [reply changelogs/unreleased/gh-6253-net-box-iproto-id.md:1]: 1 and 3 fixed. As for 2, I prefer to leave 'protocol' in the names for consistency with `required_protocol_version` and `peer_protocol_version`. We can't drop 'protocol' from the latter, because otherwise one could confuse it with `peer_version_id`, which stores Tarantool version of the server.

File: changelogs/unreleased/gh-6253-net-box-iproto-id.md (hunk 8 above)
Comment (locker) [reply changelogs/unreleased/gh-6253-net-box-iproto-id.md:2]: There's a number of reasons why explicitly requiring and reporting peer features on connect can be useful:
1. Testing. Seeing peer features in net.box is the only convenient way to test `IPROTO_ID` until more complicated features are introduced.
2. Early bail out. True, `netbox:begin()` will raise an error if transactions are not supported by the server, but a client may do some time-consuming work before they issue the failing command. It'd be more efficient to bail out as soon as we detect that a feature required for application to function properly is not supported. Of course, a client could invoke all commands they are going to use before starting to actually doing anything, but this wouldn't be really user-friendly.
3. Probing commands may increment the error counter or spam the log on the server. Requiring features on connect is more graceful.
4. There may be features that may affect the application behavior without introducing new commands to the protocol. For example, consider error marshaling, which is added to the feature set by https://github.com/tarantool/tarantool/pull/6435. Instead of handling both string errors and error objects in their application, a client may require the server to support error marshaling and refuse to work if it doesn't. This way, the client can write their application assuming that errors returned by functions called on the server are complex objects.

File: src/box/xrow.c (hunk 5 above)
Comment (Gerold103) [reply src/box/xrow.c:458]: This comment still awaits a response.

File: src/box/xrow.c [hunk 12]
```diff
@@ -903,6 +936,53 @@ xrow_encode_dml(const struct request *request, struct region *region,
 	return iovcnt;
 }
 
+int
+xrow_decode_id(struct xrow_header *row, struct id_request *request)
```
Comment (Gerold103) [inline src/box/xrow.c:961]: `row` can be made const.

File: test/box/net.box_iproto_id_gh-6253.result (hunk 10 above)
Comment (Gerold103) [reply test/box/net.box_iproto_id_gh-6253.result:1]: Look at SOP, section "Writing tests":
```
Every patch should have a test case. 

When writing tests for a feature, set some time apart to design a test plan. A good test plan lists all corner cases, takes into account how a new feature could interact with existing ones.

Select a proper location for your test: choose tap or result/reject style suite, depending on your preferences, and choose a suite according to the domain you’re working on.

We do not create a separate file for each feature - that would lead to a huge number of files.

For a bug fix, please create a separate file with an autonomous test case in a corresponding test suite. Name should follow this format:
<suite name>/gh-XXX-two-three-word-title
```
So for features we do not create a new file unless it is something big and independent, and the rule of `gh-####` applies only to bugfixes. The test you are referring to was made and pushed without a proper review, so we have a little inconsistency there.

The same can be found here: https://github.com/tarantool/tarantool/wiki/Code-review-procedure#testing.

**Review by Gerold103 — CHANGES_REQUESTED** [review]

File: src/box/xrow.c (hunk 5 above)
Comment (locker) [reply src/box/xrow.c:458]: [sorry, missed it - the comment is somehow invisible in the GitHub conversations view]

Sure, you're right. Fixed.

File: src/box/xrow.c (hunk 12 above)
Comment (locker) [reply src/box/xrow.c:961]: Fixed.

File: test/box/net.box_iproto_id_gh-6253.result (hunk 10 above)
Comment (locker) [reply test/box/net.box_iproto_id_gh-6253.result:1]: Right, renamed.

**Review by Gerold103 — APPROVED** [review]
